### PR TITLE
[DOC]: Narrative docs for PSF Photometry

### PR DESCRIPTION
**Pull request by mirca**
Hi all,

I'm writing narrative docs for the PSF photometry module.

Is there anything else that you would like to see? @hamogu @bsipocz @eteq @larrybradley 

PS: the doctests depends on #385

### CODE REVIEW

File: docs/photutils/psf.rst [hunk 1]
```diff
@@ -28,20 +28,383 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+                 make_noise_image(tshape, type='gaussian', mean=0.,
+                                  stddev=2., random_state=1))
+    >>> plt.imshow(image)
+    >>> plt.title('Simulated data')
+    >>> plt.colorbar(orientation='horizontal', fraction=0.046, pad=0.04)
+    >>> plt.show()
+
+
+.. plot::
+
```
Comment (bsipocz) [inline docs/photutils/psf.rst]: you can add ``:include-source:``, then there is no need to repeat the code above, just have it once with the ``.. plot::`` directive

**Comment by eteq** [comment]
Somewhere in these narrative docs we should mention the philosophical intent that data about astronomical sources always passes between stages as `Table`'s, and the column names should be unique so they can be passed around. (This was prompted by some discussion during the telecon)

File: docs/photutils/psf.rst (hunk 1 above)
Comment (mirca) [inline docs/photutils/psf.rst]: @bsipocz thanks!

**Comment by mirca** [comment]
@eteq I added a note about that, sounds good?

**Comment by mirca** [comment]
@eteq Thoughts?

**Comment by mirca** [comment]
@eteq tests should pass once #409 gets in

File: docs/photutils/psf.rst [hunk 2]
```diff
@@ -28,10 +28,391 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+    rcParams['font.size'] = 14
+    
+    plt.imshow(image)
+    plt.title('Simulated data')
+    plt.colorbar(orientation='horizontal', fraction=0.046, pad=0.04)
+
+
+Then let's import the required parts to set up a `~photutils.psf.DAOPhotPSFPhotometry` object:
+
+.. doctest-skip::
```
Comment (bsipocz) [inline docs/photutils/psf.rst]: remind me please, why do we skip the doctests here and below?

File: docs/photutils/psf.rst [hunk 3]
```diff
@@ -28,10 +28,391 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+
+First let's create an image with four overlapping stars:
+    
+.. plot::
+    :include-source:
+    
+    from astropy.table import Table 
+    from photutils.datasets import make_random_gaussians
+    from photutils.datasets import make_noise_image
+    from photutils.datasets import make_gaussian_sources
```
Comment (bsipocz) [inline docs/photutils/psf.rst]: these last three imports can be written in one

File: docs/photutils/psf.rst [hunk 4]
```diff
@@ -28,10 +28,391 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+    plt.colorbar(orientation='horizontal', fraction=0.046, pad=0.04)
+
+
+Then let's import the required parts to set up a `~photutils.psf.DAOPhotPSFPhotometry` object:
+
+.. doctest-skip::
+
+    >>> from photutils.detection import IRAFStarFinder
+    >>> from photutils.psf import IntegratedGaussianPRF, DAOGroup
+    >>> from photutils.background import MMMBackground
```
Comment (bsipocz) [inline docs/photutils/psf.rst]: import ``MADStdBackgroundRMS`` here instead of a new line

File: docs/photutils/psf.rst (hunk 2 above)
Comment (mirca) [inline docs/photutils/psf.rst]: No need to skip doctests here and there. Thanks for noticing!

File: docs/photutils/psf.rst [hunk 5]
```diff
@@ -28,10 +28,391 @@ model-fitting analysis, regardless to exactly what kind of model is actually
 being fit.  We take this road, using "PSF photometry" as shorthand for the
 general approach.
 
+PSF Photometry in Crowded Fields
+--------------------------------
+
+Photutils provides an implementation of the DAOPHOT algorithm
+(`~photutils.psf.DAOPhotPSFPhotometry`) proposed by Stetson in his
```
Comment (bsipocz) [inline docs/photutils/psf.rst]: could you please add a hyperlink for the paper? http://adsabs.harvard.edu/abs/1987PASP...99..191S 
Maybe for the whole ``Stetson in...paper`` part?

File: docs/photutils/psf.rst [hunk 6]
```diff
@@ -28,10 +28,391 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+    ``x_0``, ``y_0``, ``flux_0`` for the initial guesses.
+    Although this convention implies that the columns have to be renamed
+    along the process, it has the advantage of clarity so that one can
+    keep track and easily differentiate input/outputs.
+
+
+Basic Usage
+^^^^^^^^^^^
+
+The DAOPhotPSFPhotometry is the core class to implement the DAOPHOT algorithm
```
Comment (bsipocz) [inline docs/photutils/psf.rst]: May be it would be useful to do add the API links for all occurrence of the class names?

**Comment by bsipocz** [comment]
@mirca - This documentation looks really good, thank you. I've left a few minor comments, none of them is a blocker except it would be rather useful to have the doctest run whereever the can (so only skip the plotting parts).

**Comment by eteq** [comment]
@mirca - first glance this is looking pretty solid - now that #404 and #409 are in and you efficiently rebased, I'll look over this ASAP and merge it if it's all set (although probably it'll have to be this afternoon...)

File: docs/photutils/psf.rst [hunk 7]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+iterations is reached.
+
+Basically, DAOPHOT works as follows. The first step is to estimate the sky
+background. For this task, photutils provides several classes to compute
+scalar and 2D backgrounds, see `~photutils.background` for details. The next
+step is to find an initial estimate of the positions of potential sources.
+This can be accomplished by using source detection algorithms,
+which are implemented in `~photutils.detection`.
+
+After finding sources one would apply a clustering algorithm in order to label
```
Comment (hamogu) [inline docs/photutils/psf.rst]: Maybe one sentence here why it's useful to group them and not just put everything into a single group.

File: docs/photutils/psf.rst [hunk 8]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+    ...                                      fitter=my_fitter, niters=1,
+    ...                                      fitshape=(7,7))
+    >>> # get photometry results
+    >>> photometry_results, residual_image = my_photometry(image=my_image)
+
+Where ``my_finder``, ``my_group``, and ``my_bkg`` may be any suitable callable
+function. This approach allows one to customize every part of the photometry
+process provided that their input/output are compatible with the input/ouput
+expected by `~photutils.psf.DAOPhotPSFPhotometry` . See the API documentation
+for details.
```
Comment (hamogu) [inline docs/photutils/psf.rst]: API documentation coudl be a link like this: see 
```
`~photutils.psf.DAOPhotPSFPhotometry`
```
for details.

File: docs/photutils/psf.rst [hunk 9]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+                              stddev=2., random_state=1))
+
+    from matplotlib import rcParams
+    import matplotlib.pyplot as plt
+    rcParams['image.cmap'] = 'viridis'
+    rcParams['image.aspect'] = 1  # to get images with square pixels
+    rcParams['image.interpolation'] = 'nearest'
+    rcParams['image.origin'] = 'lower'
+    rcParams['font.size'] = 16
+
```
Comment (hamogu) [inline docs/photutils/psf.rst]: Can we set `rcParams` here easily or will that change all other plots in the docs, too? I don't know how matplotlib is set up in the plot directive. Does it load the deafult rcParams for every plot or does it carry them over to other plots?
Does anyone know? All those settings are sensible, be we want to be careful not to mess up things between plots.

File: docs/photutils/psf.rst [hunk 10]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+    >>> iraffind = IRAFStarFinder(threshold=3.5*std,
+    ...                           fwhm=sigma_psf*gaussian_sigma_to_fwhm,
+    ...                           minsep_fwhm=0.01, roundhi=5.0, roundlo=-5.0,
+    ...                           sharplo=0.0, sharphi=2.0) # doctest: +SKIP
+    >>> daogroup = DAOGroup(2.0*sigma_psf*gaussian_sigma_to_fwhm)
+    >>> mmm_bkg = MMMBackground()
+    >>> fitter = LevMarLSQFitter()
+    >>> psf_model = IntegratedGaussianPRF(sigma=sigma_psf) # doctest: +SKIP
+
+Note that the parameters valeus for the finder class, i.e.,
```
Comment (hamogu) [inline docs/photutils/psf.rst]: typo: values

File: docs/photutils/psf.rst [hunk 11]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+    >>> from astropy.stats import gaussian_sigma_to_fwhm
+
+Let's then instantiate the objects::
+
+    >>> bkgrms = MADStdBackgroundRMS()
+    >>> std = bkgrms(image) # doctest: +SKIP
+    >>> iraffind = IRAFStarFinder(threshold=3.5*std,
+    ...                           fwhm=sigma_psf*gaussian_sigma_to_fwhm,
+    ...                           minsep_fwhm=0.01, roundhi=5.0, roundlo=-5.0,
+    ...                           sharplo=0.0, sharphi=2.0) # doctest: +SKIP
```
Comment (hamogu) [inline docs/photutils/psf.rst]: Is the "SKIP" really needed here? As @bsipocz it's always better to run as much as we can as a doctest, because somebody will change things at some point in the future and then we will forget to update the docs...

File: docs/photutils/psf.rst [hunk 12]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+
+    >>> from photutils.psf import DAOPhotPSFPhotometry
+    >>> daophot_photometry = DAOPhotPSFPhotometry(finder=iraffind,
+    ...                                           group_maker=daogroup,
+    ...                                           bkg_estimator=mmm_bkg,
+    ...                                           psf_model=psf_model,
+    ...                                           fitter=LevMarLSQFitter(),
+    ...                                           niters=1, fitshape=(11,11)) # doctest: +SKIP
+
+As mention before, one can use the ``daophot_photometry`` object as a function
```
Comment (hamogu) [inline docs/photutils/psf.rst]: mention -> mentioned

File: docs/photutils/psf.rst [hunk 13]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+``psf_model = IntegratedGaussianPRF(sigma=sigma_psf)``:
+
+.. doctest-skip::
+
+    >>> psf_model.x_0.fixed = True
+    >>> psf_model.y_0.fixed = True
+    >>> pos = Table(names=['x_0', 'y_0'], data=[sources['x_mean'],
+    ...                                         sources['y_mean']])
+
+Note that we do not need to set the ``find`` and ``niters`` attributes in
```
Comment (hamogu) [inline docs/photutils/psf.rst]: find is called finder now.

**Comment by hamogu** [comment]
Fantastic. I left a few very small comments which will take < 3 min to implement (or ignore).

File: docs/photutils/psf.rst (hunk 9 above)
Comment (larrybradley) [inline docs/photutils/psf.rst]: @mirca Is there some reason you aren't using these settings as keywords, e.g.:
`plt.imshow(image, cmap='viridis', interpolation='nearest', origin='lower')`.

Comment (mirca) [inline docs/photutils/psf.rst]: No, there isn't. I'm going to change this, thanks :)

**Comment by mirca** [comment]
@hamogu thanks Moritz!

File: docs/photutils/psf.rst (hunk 11 above)
Comment (mirca) [inline docs/photutils/psf.rst]: @hamogu this skip is needed because ``image`` is defined outside in the ``plot`` directive right above (and the variable ``std`` depends on ``image``). Should I repeat ``image`` definition here and get rid of the doctest skips? @bsipocz

File: docs/photutils/psf.rst [hunk 14]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
 being fit.  We take this road, using "PSF photometry" as shorthand for the
 general approach.
 
+PSF Photometry in Crowded Fields
+--------------------------------
+
+Photutils provides an implementation of the DAOPHOT algorithm
```
Comment (eteq) [inline docs/photutils/psf.rst]: I would slightly rephrase this to say something like "Photutils provides a modular set of tools to perform PSF photometry for different science cases.  These are implemented as separate classes to do sub-tasks of PSF photometry.  It also provides high-level classes that connect these pieces together.  In particular, it contains an implementation of the DAOPHOT..." and then continue as is.

That is, get across the idea that it's not *just* a re-implementation of DAOPHOT, but a pluggable thing that be default is basically the DAOPHOT loop, but can be much more if you play around with it.

File: docs/photutils/psf.rst [hunk 15]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+(`~photutils.psf.DAOPhotPSFPhotometry`) proposed by `Stetson in his
+seminal paper <http://adsabs.harvard.edu/abs/1987PASP...99..191S >`_ for 
+crowded-field stellar photometry.
+The DAOPHOT algorithm consists in applying the loop FIND, GROUP, NSTAR,
+SUBTRACT, FIND until no more stars are detected or a given number of
+iterations is reached.
+
+Basically, DAOPHOT works as follows. The first step is to estimate the sky
```
Comment (eteq) [inline docs/photutils/psf.rst]: Maybe `DAOPHOT`-> here to `` `~photutils.psf.DAOPhotPSFPhotometry` `` so that it's clear this is referncing how the photutils version works?

File: docs/photutils/psf.rst [hunk 16]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+
+The `~photutils.psf.DAOPhotPSFPhotometry` is the core class to implement the
+DAOPHOT algorithm to perform PSF photometry in crowded fields.
+
+It basically encapsulates the loop "FIND, GROUP, NSTAR, SUBTRACT" in one place
+so that one can easily perform PSF photometry just by setting up a
+`~photutils.psf.DAOPhotPSFPhotometry` object.
+
+This class was implemented in such a way that it can be used as a callable
+function. The basic idea is illustrated as follows:
```
Comment (eteq) [inline docs/photutils/psf.rst]: Maybe also just mention briefly here that the "guts" of it are in `do_photometry`.  This is a useful thing because otherwise novices might not understand why e.g. the docstring of `__call__` doesn't have any of the details of how to use it.

File: docs/photutils/psf.rst (hunk 11 above)
Comment (bsipocz) [inline docs/photutils/psf.rst]: @mirca - yes, in that case I think it's better to have the duplicate and have the doctest run.

File: docs/photutils/psf.rst [hunk 17]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+Where ``my_finder``, ``my_group``, and ``my_bkg`` may be any suitable callable
+function. This approach allows one to customize every part of the photometry
+process provided that their input/output are compatible with the input/ouput
+expected by `~photutils.psf.DAOPhotPSFPhotometry` . See the API documentation
+for details.
+
+Performing PSF Photometry
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Let's take a look in a simple example with simulated stars whose PSF is
```
Comment (eteq) [inline docs/photutils/psf.rst]: "in" -> "at"

File: docs/photutils/psf.rst [hunk 18]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+    rcParams['image.interpolation'] = 'nearest'
+    rcParams['image.origin'] = 'lower'
+    rcParams['font.size'] = 16
+    
+    plt.imshow(image)
+    plt.title('Simulated data')
+    plt.colorbar(orientation='horizontal', fraction=0.046, pad=0.04)
+
+
+Then let's import the required parts to set up a `~photutils.psf.DAOPhotPSFPhotometry` object::
```
Comment (eteq) [inline docs/photutils/psf.rst]: "parts" -> "classes"

File: docs/photutils/psf.rst (hunk 11 above)
Comment (eteq) [inline docs/photutils/psf.rst]: definitely we *do* want to define the image so that it appears inline.  We sometimes repeat things with plots in this way - it is awakward but it seems to be the best way to make sure all the testing gets done...

One way to work around this in the docs themselves is to show the plots without ``:include-source:``, but in the right position.  That makes the final rendered docs look *like* you ran them in the ``>>>`` sections even though it's actually the ``.. plot::`` directive that does it.

File: docs/photutils/psf.rst (hunk 12 above)
Comment (eteq) [inline docs/photutils/psf.rst]: "one can use" -> "the way to actually *do* the photometry" or something like that.  That is, get across the idea that it's not an option, but rather the actually most important part of the whole process - it's where the photometric measurements actulally happen.

File: docs/photutils/psf.rst [hunk 19]
```diff
@@ -28,10 +28,379 @@ model-fitting analysis, regardless to exactly what kind of model is actually
...
+
+    plt.subplot(1, 2, 1)
+    plt.imshow(image)
+    plt.title('Simulated data')
+    plt.colorbar(orientation='horizontal', fraction=0.046, pad=0.04)
+    plt.subplot(1 ,2, 2)
+    plt.imshow(residual_image)
+    plt.title('Residual Image')
+    plt.colorbar(orientation='horizontal', fraction=0.046, pad=0.04)    
+
```
Comment (eteq) [inline docs/photutils/psf.rst]: Maybe a concluding sentence here along the lines of "hey look, it works!"

**Comment by eteq** [comment]
OK, finished my review @mirca.  Overall looks great.  The one possibly complicated thing is that I suggested a possible fix to the need for the skipping (which we want to avoid if at all possible) by not including the code for the plots in the rendered docs.  Instead you essentially repeat it twice in the RST, but then sphinx will hide the source for the docs, and if you're clever about placement it looks *like* it came from the code in the ``>>>`` blocks.

**Comment by bsipocz** [comment]
@mirca - You need to use the ``doctest-requires`` directive to make this pass when no scipy and matplotlib available. Adding ``.. doctest-requires:: scipy, matplotlib`` before the script should make it pass on the remaining travis builds.

**Comment by eteq** [comment]
Alright, this is pretty much looking all set.  So I'm merging this now with a few minor wording changes I found easier to just make during the merge.  Thanks @mirca!
